lib/discrete_maths/logic_operations.py

```diff
@@ -0,0 +1,178 @@
+
+from lib.parsers import shunting_yard
+
+# Operator Precedence Dictionary - = == iff, > == implies, v == or, ^ == and, ~ == not
+OP_PRECIDENCE = {"=": 2, ">": 3, "v": 4, "^": 5, "~": 6}
+
+def negation(a):
+    return not a
+
+def conjunction(a, b):
+    return a and b
+
+def disjunction(a, b):
+    return a or b
+
+def implies(a, b):
+    return (not a) or b
+
+def iff(a, b):
+    return (a and b) or ((not a) and (not b))
+
+def xor(a, b):
+    return (a and (not b)) or ((not a) and b)
+
+def nand(a, b):
+    return not (a and b)
+
+def nor(a, b):
+    return not (a or b)
+
+def xnor(a, b):
+    return (not a and not b) or (a and b)
+
+# Check if a proposition is balanced
+def is_balanced_proposition(proposition: str) -> bool:
+    balance = 0
+    for char in proposition:
+        if char == '(':
+            balance += 1
+        elif char == ')':
+            balance -= 1
+        if balance < 0:
+            return False
+    
+    return balance == 0
+
+# Check if a proposition is well-formed - no 2 consecutive operators or variables
+def is_well_formed_proposition(proposition: str) -> bool:
+    for i in range(len(proposition) - 1):
+        if proposition[i] in ['v', '^', '>', '='] and proposition[i + 1] in ['v', '^', '>', '=']:
+            return False
+        if proposition[i].isalpha() and proposition[i + 1].isalpha():
+            return False
+        
+    return True
+
+# Check if a proposition is valid
+def is_valid_proposition(proposition: str) -> bool:
+    if not proposition:
+        raise SyntaxError("Empty proposition")
+
+    for char in proposition:
+        if char not in ['~', '(', ')', 'v', '^', '>', '=', ' ']:
+            if not char.isalpha():
+                raise ValueError("Invalid character in proposition: " + char)
+    
+    if not is_balanced_proposition(proposition):
+        raise SyntaxError("Unbalanced proposition")
+    
+    if not is_well_formed_proposition(proposition):
+        raise SyntaxError("Proposition is not well-formed")
+    
+    return True
+    
+# Calculate the result of a proposition given a set of variable values and an RPN version of the proposition
+def calculate_rpn_prop(rpn: list, variable_values: dict) -> bool:
+    stack = []
+
+    for token in rpn:
+        if token in variable_values.keys():
+            stack.append(variable_values[token])
+        elif token == "~":
+            stack.append(negation(stack.pop()))
+        elif token == "^":
+            stack.append(conjunction(stack.pop(), stack.pop()))
+        elif token == "v":
+            stack.append(disjunction(stack.pop(), stack.pop()))
+        elif token == ">":
+            stack.append(implies(stack.pop(), stack.pop()))
+        elif token == "=":
+            stack.append(iff(stack.pop(), stack.pop()))
+    
+    return stack.pop()
+
+# Create a list of lists of all the possible values for the variables in the rpn proposition
+def create_variable_values(rpn: list) -> list:
```

Comment:
I think `create_values_for_variables` or `generate_values_for_variables` would be more meaningful and would allow you to not have to comment it and explain it.  

This is a function to create all combination of values? You can simplify this a lot. For example you can use regex to extract all variables. Also, you can do part of the process directly while creating the truth table instead of doing the process of creating a whole list of combination. It's much easier and you're doing everything at the same time instead of doing it multiple time. 

Also, see comment for `create_truth_table`. This function is part of it.